File: 2023/day4/day-2.py
import sys
import pathlib
import math

# récupération de l'input ou du test
filename = "test-2"
if len(sys.argv)>=2 :
    filename = sys.argv[1]

file = open(str(pathlib.Path(__file__).parent.resolve())+"/"+filename,'r')
lines = file.readlines();

def debug(string):
    string=string
    print(string)

tab = []
cardCopyTab = []


for line in lines:
    line=line.strip()
    debug(line)
    tab.append(line)
    cardCopyTab.append(1)

#Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53

def getCardNumber(str):
    return str[:str.find(":")][str.find(" "):].strip()

def getWiningNumbers(str):
    return str[:str.find("|")][str.find(":")+1:].strip()

def getPlayingNumbers(str):
    return str[str.find("|")+1:].strip()

print(str(cardCopyTab))
total = 0
for y in range(len(tab)):
    cardMatch = 0
    for curNum in getPlayingNumbers(tab[y]).split(" "):
        if len(curNum) == 0:
            continue
        if (" "+curNum+" ") in " "+getWiningNumbers(tab[y])+" ":
            cardMatch+=1
    for i in range(cardMatch):
        cardCopyTab[y+i+1]+=cardCopyTab[y]
    debug( getCardNumber(tab[y]) + "==>" + getWiningNumbers(tab[y]) + "==>" + getPlayingNumbers(tab[y]) +" ====== " + str(cardMatch))
    print(str(cardCopyTab))


total = 0
for i in cardCopyTab:
    total += i

print(str(cardCopyTab))
print(total)

File: 2024/day4/day-2.js
const fs = require('fs');

fs.readFile('input', 'utf8', (err, data) => {
  if (err) {
    console.error(err);
    return;
  }

  let total = 0;
  const lines = data.replace(/ /g,'').split('\n');
  for( let j = 1; j < lines.length-1; j++){
      for( let i = 1; i < lines[j].length-1; i++){
        if( lines[j][i]=='A'){
          let str=j+":"+i;
          if( ((lines[j-1][i-1]==='M' && lines[j+1][i+1]==='S') || (lines[j-1][i-1]==='S' && lines[j+1][i+1]==='M')) &&
              ((lines[j+1][i-1]==='M' && lines[j-1][i+1]==='S') || (lines[j+1][i-1]==='S' && lines[j-1][i+1]==='M'))){
            total++;
            str+="X";
          }
          console.log(str);
        }

      }
  };

  console.log("Total: "+total);
});